### DISCUSSION TITLE
`Query.update` events, what's the right way to get and work with them

### DISCUSSION THREAD

**Discussion by nicmollel-ro** [sqlalchemy/sqlalchemy#8497] [Usage Questions] [answered]
hello, i am working in an application where updates are primarily through helper functions using `Query.update`. I want to generate an update log out of the database for some model and I came up with simplified use case. At the end, I want to know 
- when a resource i care about is updated 
- pass the latest resource state to another function. I will not be going any resource modification at that stage. 

Is this a right approach? How do I get it to actually not close the connection on the `CursorResult`

```python
from sqlalchemy import create_engine
from sqlalchemy import Column
from sqlalchemy import event
from sqlalchemy import Integer
from sqlalchemy import String
from sqlalchemy.orm import declarative_base
from sqlalchemy.orm import Session
from sqlalchemy.orm import sessionmaker

Base = declarative_base()


class Resource(Base):
    __tablename__ = "resource_table"
    id = Column(Integer, primary_key=True)
    data = Column(String)


engine = create_engine("sqlite://", echo=True)
Base.metadata.create_all(engine)

LocalSession = sessionmaker(engine)


@event.listens_for(Session, "after_bulk_update")
def event_handler(update_context):
    print("INFO: in event handler")
    for record in update_context.result.scalars():
        print(f"INFO: got record with id: {record.id}")


def update_resource(*, session: Session, id: int, **update_args) -> int:
    rowcount = session.query(Resource).filter_by(id=id).update(update_args, synchronize_session="fetch")
    print("INFO: update, before commit")
    session.commit()

    return rowcount


print("INFO: start")

r_id = None
with LocalSession() as s:
    print("INFO: create resource")
    r = Resource(data="init data")
    s.add(r)
    s.commit()
    r_id = r.id

data_update = "updated details"
with LocalSession() as s:
    c = update_resource(session=s, id=r_id, **{"data": data_update})

    assert c == 1
    rs = s.query(Resource).get(r_id)
    assert rs.data == data_update

```

current error
```text
ResourceClosedError: This result object does not return rows. It has been closed automatically.
```

**Comment by zzzeek** [comment, marked as answer]
an UPDATE statement does not return any rows unless you use RETURNING.

The legacy `query.update()` doesn't have any RETURNING support.   You can use returning explicitly as documented at https://docs.sqlalchemy.org/en/14/orm/persistence_techniques.html#orm-dml-returning-objects .     I'm not sure what interaction this would have with the "after_bulk_update" event, you can try it out.  otherwise just do the event thing you want to do inside of your update_resource() function, not sure an event is really needed here.

**Reply by nicmollel-ro** [reply]
> You can use returning explicitly as documented at https://docs.sqlalchemy.org/en/14/orm/persistence_techniques.html#orm-dml-returning-objects . I'm not sure what interaction this would have with the "after_bulk_update" event, you can try it out.

i tried this out on postgresql, it did not cause any event firing for `after_bulk_update`. I am going to eschew the event and do the change log tracking in my update for now

**Reply by zzzeek** [reply]
the new style update() does not interact with after_bulk_update, I came across this today.   that event is "legacy" in 2.0 as it only applies to Query.

to intercept all ORM queries in 1.4 and 2.0 you can use [do_orm_execute](https://docs.sqlalchemy.org/en/14/orm/events.html?highlight=do_orm_execute#sqlalchemy.orm.SessionEvents.do_orm_execute).  Every query emitted on behalf of the ORM, not including those within the flush() process, should be going through this event hook.
